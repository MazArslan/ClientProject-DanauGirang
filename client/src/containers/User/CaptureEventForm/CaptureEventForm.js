import React, { useState, useEffect, useCallback } from "react";
import {
  Container,
  Row,
  Col,
  ButtonGroup,
  Button,
  Form,
  FormGroup,
  Label,
  Input
} from "reactstrap";
import propTypes from "prop-types";

import useHttp from "../../../hooks/http";
import { updateObject } from "../../../shared/utility";
import Spinner from "../../../components/UI/Spinner/Spinner";
import Image from "../../../components/UI/Image/Image";
import AIImage from "../../../components/UI/AIImage/AIImage";
import RadioGroup from "../../../components/UI/RadioGroup/RadioGroup";
import ConfirmationModal from "../../../components/UI/ConfirmationModal/ConfirmationModal";

/**
 * Capture event form component
 * The form for the user to add information about a capture event
 * @param { tags, captureEventId, submitted } props
 */
const CaptureEventForm = props => {
  const { tags, captureEventId, submitted } = props;

  const [modal, setModal] = useState(false);
  const [capturedImage, setCapturedImage] = useState();
  const [aiImages, setAIImages] = useState();
  const { isLoading, error, data, sendRequest } = useHttp();

  // Form input state
  const [radioButtonForm, setRadioButtonForm] = useState({});
  const [checkboxes, setCheckboxes] = useState([]);
  const [textAreaForm, setTextAreaForm] = useState("");

  useEffect(() => {
    // Retrive images for the current card
    // If statement prevents bug where request is made with NaN values
    if (captureEventId !== undefined) {
      sendRequest(
        `http://localhost:3000/api/captureEvent/${captureEventId}/form`,
        "GET"
      );
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [captureEventId]);

  useEffect(() => {
    // Add the retrieved images into state
    if (data !== null && !data.error && data.capturedImage) {
      setCapturedImage(data.capturedImage);
      setAIImages(data.aiImages);
      if (data.captureEvent !== null) {
        setTextAreaForm(data.captureEvent.notes);
        setCheckboxes(data.captureEvent.tags);
      }
    }
  }, [data]);

  const inputChangedHandler = event => {
    setTextAreaForm(event.target.value);
  };

  const clearAllFormData = () => {
    setRadioButtonForm({});
    setCheckboxes([]);
    setTextAreaForm("");
  };

  // Triggered when image card submit button clicked
  const cardSubmissionHandler = e => {
    e.preventDefault();
    setModal(true);
  };

  const addImageCardHandler = useCallback(
    (eventId, aiCheck, assignedTags, text) => {
      const body = {};
      body.aiCheck = aiCheck;
      body.tags = assignedTags;
      body.notes = text;
      sendRequest(
        `http://localhost:3000/api/captureEvent/${eventId}/form`,
        "POST",
        JSON.stringify(body)
      );
    },
    [sendRequest]
  );

  const form = (
    <Form onSubmit={cardSubmissionHandler} className="textarea-form">
      <Row>
        <Col xs="2" />
        <Col className="text-center" lg="8">
          <FormGroup>
            <Label for="notesText">Enter additional notes:</Label>
            <Input
              type="textarea"
              name="text"
              id="notesText"
              value={textAreaForm}
              onChange={e => inputChangedHandler(e)}
            />
          </FormGroup>
        </Col>
        <Col xs="2" />
      </Row>
      <Button className="submit-button" name="submit" color="success">
        Submit
      </Button>
    </Form>
  );

  const onCheckboxBtnClick = selected => {
    const index = checkboxes.indexOf(selected);
    if (index < 0) {
      checkboxes.push(selected);
    } else {
      checkboxes.splice(index, 1);
    }
    setCheckboxes([...checkboxes]);
  };

  const tagsForm = (
    <React.Fragment>
      <Col className="text-center">
        <ButtonGroup>
          {tags.map(tag => (
            <Button
              outline
              key={`${tag}`}
              color="primary"
              onClick={() => onCheckboxBtnClick(`${tag}`)}
              active={checkboxes.includes(`${tag}`)}
            >
              {tag}
            </Button>
          ))}
        </ButtonGroup>
      </Col>
    </React.Fragment>
  );

  let originalImage = null;
  if (capturedImage !== undefined) {
    originalImage = (
      <React.Fragment>
        <Image filePath={capturedImage.filePath} id={capturedImage._id} />
      </React.Fragment>
    );
  }

  const handleRadioElementChange = useCallback(
    (e, id) => {
      const updatedRadioForm = updateObject(radioButtonForm, {
        [id]: e.target.value
      });
      setRadioButtonForm(updatedRadioForm);
    },
    [radioButtonForm]
  );

  // let imagesForm = error ? <p>Images cannot be loaded</p> : <Spinner />;
  let imagesForm = error ? <p>Images cannot be loaded</p> : <Spinner />;
  if (aiImages && !isLoading) {
    imagesForm = (
      <React.Fragment>
        {aiImages.map(image => {
          let aiTitle = null;
          if (image.type === "MICROSOFT_AI") {
            aiTitle = (
              <Row>
                <Col className="text-center">
                  <p>
                    <b>Microsoft MegaDetector AI image</b>
                  </p>
                </Col>
              </Row>
            );
          }
          return (
            <div key={image._id}>
              {aiTitle}
              <AIImage
                filePath={image.filePath}
                detections={image.detections}
              />
              <Row>
                <Col className="text-center">
                  <p>Has the AI correctly identified all of the animals?</p>
                </Col>
              </Row>
              <RadioGroup
                aiId={image._id}
                radioButtonForm={radioButtonForm}
                selectedButtonChanged={e =>
                  handleRadioElementChange(e, image._id)
                }
              />
            </div>
          );
        })}
      </React.Fragment>
    );
  }

  const formSubmissionHandler = e => {
    e.preventDefault();
    addImageCardHandler(
      captureEventId,
      radioButtonForm,
      checkboxes,
      textAreaForm
    );
    if (submitted != null) {
      submitted(e);
    }
    clearAllFormData();
    setModal(false);
  };

  // Confirmation modal toggle handler
  const toggleModal = useCallback(
    e => {
      e.preventDefault();
      setModal(!modal);
    },
    [modal]
  );

  let confirmationModal = null;
  if (modal) {
    confirmationModal = (
      <ConfirmationModal
        modal={modal}
        toggle={toggleModal}
        onSubmit={e => formSubmissionHandler(e)}
      />
    );
  }

  return (
    <React.Fragment>
      <Container>
        {confirmationModal}
        <div className="capture-event-form">
          {originalImage}
          {imagesForm}
          <Row>
            <Col className="text-center">
              <p>Select appropriate tags for the image:</p>
            </Col>
          </Row>
          <Row>{tagsForm}</Row>
          <Row>
            <Col className="text-center">{form}</Col>
          </Row>
        </div>
      </Container>
    </React.Fragment>
  );
};

CaptureEventForm.propTypes = {
  tags: propTypes.array.isRequired,
  captureEventId: propTypes.string.isRequired,
  submitted: propTypes.func
};

export default CaptureEventForm;
