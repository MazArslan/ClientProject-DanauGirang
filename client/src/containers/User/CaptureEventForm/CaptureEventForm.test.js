import React from "react";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";
import CaptureEventForm from "./CaptureEventForm";

Enzyme.configure({ adapter: new Adapter() });

describe("<CaptureEventForm/>", () => {
  const onActionMock = jest.fn();
  const props = {
    captureEventId: "CaptureEventId1",
    tags: ["Tiger", "Elephant", "Tapir"],
    submitted: onActionMock
  };

  describe("component rendering", () => {
    const wrapper = shallow(<CaptureEventForm {...props} />);

    it("should render shallow correctly with props", () => {
      expect(wrapper.exists()).toBe(true);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("should render tags selection element", () => {
      expect(
        wrapper
          .find("Button")
          .at(0)
          .props().children
      ).toEqual("Tiger");
      expect(
        wrapper
          .find("Button")
          .at(1)
          .props().children
      ).toEqual("Elephant");
      expect(
        wrapper
          .find("Button")
          .at(2)
          .props().children
      ).toEqual("Tapir");
    });

    it("should render notes form field", () => {
      expect(wrapper.find("Label").props().for).toEqual("notesText");
      expect(wrapper.find("Label").props().children).toEqual(
        "Enter additional notes:"
      );
      expect(wrapper.find('Input[type="textarea"]').props().id).toEqual(
        "notesText"
      );
    });

    it("should set the notes element value on change event", () => {
      expect(
        wrapper.find('Input[type="textarea"]').simulate("change", {
          target: {
            value: "Updated"
          }
        })
      );
      expect(wrapper.find('Input[type="textarea"]').props().value).toEqual(
        "Updated"
      );
    });

    it("should render submit button", () => {
      const submitButton = wrapper.find('Button[className="submit-button"]');
      expect(submitButton.props().children).toEqual("Submit");
    });
  });
});
