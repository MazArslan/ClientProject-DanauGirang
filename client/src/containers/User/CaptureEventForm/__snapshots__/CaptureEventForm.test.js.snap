// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CaptureEventForm/> component rendering should render shallow correctly with props 1`] = `
<Fragment>
  <Container
    tag="div"
  >
    <div
      className="capture-event-form"
    >
      <spinner />
      <Row
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <Col
          className="text-center"
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <p>
            Select appropriate tags for the image:
          </p>
        </Col>
      </Row>
      <Row
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <Col
          className="text-center"
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <ButtonGroup
            role="group"
            tag="div"
          >
            <Button
              active={false}
              color="primary"
              key="Tiger"
              onClick={[Function]}
              outline={true}
              tag="button"
            >
              Tiger
            </Button>
            <Button
              active={false}
              color="primary"
              key="Elephant"
              onClick={[Function]}
              outline={true}
              tag="button"
            >
              Elephant
            </Button>
            <Button
              active={false}
              color="primary"
              key="Tapir"
              onClick={[Function]}
              outline={true}
              tag="button"
            >
              Tapir
            </Button>
          </ButtonGroup>
        </Col>
      </Row>
      <Row
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <Col
          className="text-center"
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <Form
            className="textarea-form"
            onSubmit={[Function]}
            tag="form"
          >
            <Row
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <Col
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
                xs="2"
              />
              <Col
                className="text-center"
                lg="8"
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <FormGroup
                  tag="div"
                >
                  <Label
                    for="notesText"
                    tag="label"
                    widths={
                      Array [
                        "xs",
                        "sm",
                        "md",
                        "lg",
                        "xl",
                      ]
                    }
                  >
                    Enter additional notes:
                  </Label>
                  <Input
                    id="notesText"
                    name="text"
                    onChange={[Function]}
                    type="textarea"
                    value=""
                  />
                </FormGroup>
              </Col>
              <Col
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
                xs="2"
              />
            </Row>
            <Button
              className="submit-button"
              color="success"
              name="submit"
              tag="button"
            >
              Submit
            </Button>
          </Form>
        </Col>
      </Row>
    </div>
  </Container>
</Fragment>
`;
